code/datums/datum.dm

```diff
@@ -322,37 +322,51 @@
  * * priority - Priority used when sorting the filter.
  * * params - Parameters of the filter.
  */
-/datum/proc/add_filter(name, priority, list/params)
+/datum/proc/add_filter(name, priority, list/params, update = TRUE)
 	LAZYINITLIST(filter_data)
 	var/list/copied_parameters = params.Copy()
+	copied_parameters["name"] = name
 	copied_parameters["priority"] = priority
-	filter_data[name] = copied_parameters
-	update_filters()
+	for (var/list/filter_info as anything in filter_data)
+		if (filter_info["name"] == name)
+			filter_data -= filter_info
+			break
+	BINARY_INSERT_LIST(list(copied_parameters), filter_data, copied_parameters, "priority", COMPARE_KEY)
+	if (!update)
+		return
+
+	for (var/index in 1 to length(filter_data))
+		var/list/filter_info = filter_data[index]
+		if (filter_info["name"] == name)
+			update_filters(index) // Only update from the index we care about to avoid unnecessarily recreating filters
+			return
 
 ///A version of add_filter that takes a list of filters to add rather than being individual, to limit calls to update_filters().
-/datum/proc/add_filters(list/list/filters)
+/datum/proc/add_filters(list/list/filters, update = TRUE)
 	LAZYINITLIST(filter_data)
-	for(var/list/individual_filter as anything in filters)
-		var/list/params = individual_filter["params"]
-		var/list/copied_parameters = params.Copy()
-		copied_parameters["priority"] = individual_filter["priority"]
-		filter_data[individual_filter["name"]] = copied_parameters
-	update_filters()
-
-/// Reapplies all the filters.
-/datum/proc/update_filters()
+	for (var/list/individual_filter as anything in filters)
+		add_filter(individual_filter["name"], individual_filter["priority"], individual_filter, update = FALSE)
+	if (update)
+		update_filters()
+
+/// Reapplies all the filters. If start_index is passed, only a portion of all filters are reapplied starting from said index
+/datum/proc/update_filters(start_index = null)
 	ASSERT(isatom(src) || isimage(src))
 	var/atom/atom_cast = src // filters only work with images or atoms.
-	atom_cast.filters = null
-	sortTim(filter_data, GLOBAL_PROC_REF(cmp_filter_data_priority), TRUE)
-	for(var/filter_raw in filter_data)
-		var/list/data = filter_data[filter_raw]
-		var/list/arguments = data.Copy()
+	if (!start_index)
+		atom_cast.filters = null
+
+	for (var/index in start_index || 1 to length(filter_data))
+		var/list/filter_info = filter_data[index]
+		var/list/arguments = filter_info.Copy()
 		arguments -= "priority"
-		atom_cast.filters += filter(arglist(arguments))
+		if (start_index) // See https://www.byond.com/forum/post/2980598 as to why we cannot just override the existing filter
+			atom_cast.filters -= filter_info["name"] // We're trapped in the belly of this horrible machine
+		atom_cast.filters += filter(arglist(arguments)) // And the machine is bleeding to death
```

Comment:
I suspect we could get around a lot of the hell here by using a temporary list to hold filter() results, though I am also unsure of where the cost here comes from so it's possible that wouldn't help THAT much. should improve appearance hell signficantly tho